edgelet/edgelet-docker/src/runtime.rs

```diff
@@ -401,23 +402,74 @@ impl ModuleRuntime for DockerModuleRuntime {
+                        // the container VHD is help open by `System`.
+                        // Specifically, the start container API endpoint returns an error with the message:
+                        //
+                        //     hcsshim::ActivateLayer failed in Win32: The process cannot access the file because it is being used by another process. (0x20)
+                        //
+                        // The only way to resolve this is to delete the container and let it be re-created by Edge Agent.
+                        // If Edge Agent itself needs to be deleted, the watchdog will recreate it.
+                        let mut is_hcs_vhd_in_use_by_another_process = false;
+                        if cfg!(windows) {
+                            if let DockerError::Api(err) = &err {
+                                if err.code == StatusCode::INTERNAL_SERVER_ERROR {
+                                    if let Some(message) = crate::error::get_message(err) {
+                                        is_hcs_vhd_in_use_by_another_process = message.starts_with(
+                                            "hcsshim::ActivateLayer failed in Win32: The process cannot access the file because it is being used by another process. (0x20)",
+                                        );
+                                    }
+                                }
+                            }
+                        }
+
                         let err = Error::from_docker_error(
                             err,
-                            ErrorKind::RuntimeOperation(RuntimeOperation::StartModule(id)),
+                            ErrorKind::RuntimeOperation(RuntimeOperation::StartModule(id.clone())),
                         );
-                        log_failure(Level::Warn, &err);
-                        Err(err)
+
+                        if is_hcs_vhd_in_use_by_another_process {
+                            future::Either::B(client
+                                .container_api()
+                                .container_delete(
+                                    &id,
+                                    /* remove volumes */ false,
+                                    /* force */ true,
+                                    /* remove link */ false,
+                                ).then(|result| {
+                                if let Err(err) = result {
+                                    let err = Error::from_docker_error(
+                                        err,
+                                        ErrorKind::RuntimeOperation(RuntimeOperation::StartModule(id)),
+                                    );
+                                    log_failure(Level::Warn, &err);
+                                }
+
+                                // Return original error so that caller interprets the start operation as a failure
+                                Err(err)
```

Comment:
Is there a scenario where an `err` will be returned without first being logged? I ask because the call to `log-failure()` was moved into an `if` statement.

Reply:
There are two errors, one from the original `container_start`, and one from the `container_delete`. The `container_delete` error is logged here. The `container_start` error is logged in the `map_err` at the end of the combinator chain.